File: Java/2023/Solvers/Day04Solver.java
package Solvers;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import Inputs.InputParser;

public class Day04Solver extends InputParser {
    public int day04Part01() throws IOException { 
        ArrayList<String> lines = ReadLines("C:/Users/CameronGraybill/Documents/Code/AoC2023Java/Inputs/Day04Input.txt");

        int answer = 0;
        int colon = lines.get(0).indexOf(":");
        int divider = lines.get(0).indexOf("|");

        for (String line : lines) {
            List<String> winningNumbers = Arrays.asList(line.substring(colon, divider).split("\\s+"));
            List<String> cardNumbers = Arrays.asList(line.substring(divider).split("\\s+"));

            int pointCounter = 0;

            for (String wn : winningNumbers) {
                if (cardNumbers.contains(wn) && pointCounter == 0) {
                    pointCounter += 1;
                    continue;
                } else if (cardNumbers.contains(wn)) {
                    pointCounter *= 2;
                }
            }

            answer += pointCounter;
        }
        
        return answer;
    }

    public int day04Part02() throws IOException { 
        ArrayList<String> lines = ReadLines("C:/Users/CameronGraybill/Documents/Code/AoC2023Java/Inputs/Day04Input.txt");

        int answer = 0;
        int colon = lines.get(0).indexOf(":");
        int divider = lines.get(0).indexOf("|");

        for (String line : lines) {
            List<String> winningNumbers = Arrays.asList(line.substring(colon, divider).split("\\s+"));
            List<String> cardNumbers = Arrays.asList(line.substring(divider).split("\\s+"));

            int cardCounter = 1;

            for (String wn : winningNumbers) {
                if (cardNumbers.contains(wn)) {
                    cardCounter += 1;
                    continue;
                } 
            }
            answer += cardCounter;
        }
        return answer;
    }
}

File: TypeScript/src/2022/Solvers/Day04Solver.ts
import { parseToStrings } from "../Helpers/InputParser.js";

var assignments = parseToStrings("TypeScript/src/2022/Inputs/Day04Input.txt");

function day04Part01() {
    var answer = 0;

    for (const assignment of assignments) {
        var parsed = parseAssignment(assignment);
        var pairOne = parsed.firstPair;
        var pairTwo = parsed.secondPair;

        if ((pairOne[0] >= pairTwo[0] && pairOne[1] <= pairTwo[1]) || (pairTwo[0] >= pairOne[0] && pairTwo[1] <= pairOne[1])) {
            answer++;
        }
    }

    return answer;
}

function day04Part02() {
    var answer = 0;

    for (const assignment of assignments) {
        var parsed = parseAssignment(assignment);
        var pairOne = parsed.firstPair;
        var pairTwo = parsed.secondPair;

        if ((pairOne[1] >= pairTwo[0] && pairOne[0] <= pairTwo[1])) {
            answer++;
        }
    }

    return answer;
}

function parseAssignment(assignment: string) {
    var splitIndex = assignment.indexOf(",");
    var firstPair = assignment.substring(0, splitIndex).split("-").map(x => parseInt(x));
    var secondPair = assignment.substring(splitIndex + 1).split("-").map(x => parseInt(x));;
    
    return {firstPair, secondPair};
}

console.log(day04Part02());